File: dotnet/tests/Buble.Sdk.Tests/AppsTests.cs
using Xunit;

namespace Buble.Sdk.Tests;

public sealed class AppsTests
{
    [Fact]
    public async Task CreatesAppGenerationWithFlatParameters()
    {
        var handler = new FakeHttpMessageHandler();
        handler.EnqueueJson("""{"data":{"id":"app_task_1","status":"pending"}}""");
        using var client = new BubleClient(new BubleClientOptions
        {
            ApiKey = "sk_test",
            BaseUrl = "https://example.test",
            HttpClient = new HttpClient(handler)
        });

        await client.Apps.Generations.CreateAsync("video-background-remover", new Dictionary<string, object?>
        {
            ["source_video"] = new[] { "https://example.test/source.mp4" },
            ["subject_is_person"] = true
        });

        Assert.Equal("/api/v1/apps/video-background-remover/generations", handler.Requests[0].Uri.AbsolutePath);
        Assert.Contains("\"source_video\"", handler.Requests[0].Body);
        Assert.Contains("\"subject_is_person\":true", handler.Requests[0].Body);
    }
}

File: swift/Tests/BubleTests/AppsTests.swift
import Buble
import XCTest

final class AppsTests: XCTestCase {
    func testListsAppsWithQuery() async throws {
        let transport = try TestHTTPTransport(responses: [
            .json([
                "data": [
                    [
                        "id": "video-background-remover",
                        "input_parameters": [
                            ["name": "source_video", "type": "array"]
                        ]
                    ]
                ]
            ])
        ])
        let client = try makeClient(transport: transport)

        let apps = try await client.apps.list(options: ListAppsOptions(limit: 20))

        XCTAssertEqual(apps.data.first?.id, "video-background-remover")
        let requests = await transport.recordedRequests()
        XCTAssertEqual(requests.first?.url?.query, "limit=20")
    }

    func testCreatesAndWaitsForAppGeneration() async throws {
        let transport = try TestHTTPTransport(responses: [
            .json(["data": ["id": "task_1", "status": "pending"]]),
            .json([
                "data": [
                    "id": "task_1",
                    "status": "success",
                    "result": ["videos": [["url": "https://example.com/video.mp4"]]]
                ]
            ])
        ])
        let client = try makeClient(transport: transport)

        let task = try await client.apps.generations.create(
            "video-background-remover",
            body: ["source_video": ["https://example.com/source.mp4"]]
        )
        let result = try await client.apps.generations.wait(
            "video-background-remover",
            task.data.id,
            options: WaitOptions(interval: 0.001, timeout: 1)
        )

        XCTAssertEqual(result.data.result?.videos?.first?.url.absoluteString, "https://example.com/video.mp4")
        let requests = await transport.recordedRequests()
        XCTAssertEqual(requests[0].url?.path, "/api/v1/apps/video-background-remover/generations")
        XCTAssertEqual(requests[1].url?.path, "/api/v1/apps/video-background-remover/generations/task_1")
    }
}
